### PR TITLE
Upgrading HDFS Repository Plugin to use HDFS 2.8.1 Client

### LINKED ISSUE #25450: Hadoop 2.7.x is not compatible with Java 9
In the CI tests for Java 9, we are getting failures in the hdfsFixture:

```
Picked up JAVA_TOOL_OPTIONS: -Dfile.encoding=UTF8
    Exception in thread "main" java.lang.ExceptionInInitializerError
    	at org.apache.hadoop.util.StringUtils.<clinit>(StringUtils.java:80)
    	at org.apache.hadoop.security.SecurityUtil.getAuthenticationMethod(SecurityUtil.java:611)
    	at org.apache.hadoop.security.UserGroupInformation.initialize(UserGroupInformation.java:272)
    	at org.apache.hadoop.security.UserGroupInformation.setConfiguration(UserGroupInformation.java:311)
    	at hdfs.MiniHDFS.main(MiniHDFS.java:89)
    Caused by: java.lang.StringIndexOutOfBoundsException: begin 0, end 3, length 1
    	at java.base/java.lang.String.checkBoundsBeginEnd(String.java:3116)
    	at java.base/java.lang.String.substring(String.java:1885)
    	at org.apache.hadoop.util.Shell.<clinit>(Shell.java:51)
    	... 5 more
```

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+java9-periodic/3051/console

This is a known issue with Hadoop 2.7.x, https://issues.apache.org/jira/browse/HADOOP-14586, and it may not be fixed. A "hacky" workaround was used by the reported of the issue, https://git-wip-us.apache.org/repos/asf?p=lucene-solr.git;a=commitdiff;h=85a27a2. Maybe we can bring our version of the client jars up to a newer version?

### PR DESCRIPTION
**Pull request by jbaiera**
This should fix #25450. We are upgrading our version of the HDFS client used within the HDFS Repository Plugin in order to fix issues with Hadoop's client code parsing JDK9's version string. The HDFS Test fixture has also been updated to version 2.8.1.

### CODE REVIEW

File: plugins/repository-hdfs/build.gradle [hunk 1]
```diff
@@ -473,7 +470,36 @@ thirdPartyAudit.excludes = [
...
+  'com.squareup.okhttp.MediaType',
+  'com.squareup.okhttp.OkHttpClient',
+  'com.squareup.okhttp.Request$Builder',
+  'com.squareup.okhttp.RequestBody',
+  'com.squareup.okhttp.Response',
+  'com.squareup.okhttp.ResponseBody',
+
+  // HDFS
+  'io.netty.channel.ChannelOption',
+  'io.netty.util.ReferenceCountUtil'
```
Comment (jasontedor) [inline plugins/repository-hdfs/build.gradle]: Can you sort these, for example, there are already `io.netty` excludes above?

File: plugins/repository-hdfs/src/test/java/org/elasticsearch/repositories/hdfs/HdfsClientThreadLeakFilter.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ * is started only once in a class's static initialization. This currently breaks our testing as this
+ * thread leaks out of the client and is picked up by the test framework. This thread filter is meant
+ * to ignore the offending thread until a version of Hadoop is released that addresses the incorrect
+ * interrupt handling.
+ *
+ * @see <a href="https://issues.apache.org/jira/browse/HADOOP-12829">https://issues.apache.org/jira/browse/HADOOP-12829</a>
+ * @see "org.apache.hadoop.fs.FileSystem.Statistics.StatisticsDataReferenceCleaner"
+ * @see "org.apache.hadoop.fs.FileSystem.Statistics"
+ */
+public class HdfsClientThreadLeakFilter implements ThreadFilter {
```
Comment (jasontedor) [inline plugins/repository-hdfs/src/test/java/org/elasticsearch/repositories/hdfs/HdfsClientThreadLeakFilter.java]: Can it be package private and final?

File: plugins/repository-hdfs/src/test/java/org/elasticsearch/repositories/hdfs/HdfsClientThreadLeakFilter.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+package org.elasticsearch.repositories.hdfs;
+
+import com.carrotsearch.randomizedtesting.ThreadFilter;
+
+/**
+ * In Hadoop 2.8.0, there is a thread that is started by the filesystem to clean up old execution stats.
+ * This thread ignores all interrupts, catching InterruptedException, logging it, and continuing on
+ * with it's work. The thread is a daemon, so it thankfully does not stop the JVM from closing, and it
```
Comment (jasontedor) [inline plugins/repository-hdfs/src/test/java/org/elasticsearch/repositories/hdfs/HdfsClientThreadLeakFilter.java]: Nit: `it's` -> `its`

**Review by jasontedor — APPROVED** [review]
LGTM, but please address the comments that I left.

File: plugins/repository-hdfs/src/test/java/org/elasticsearch/repositories/hdfs/HdfsClientThreadLeakFilter.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ * @see "org.apache.hadoop.fs.FileSystem.Statistics"
+ */
+public class HdfsClientThreadLeakFilter implements ThreadFilter {
+
+    private static final String OFFENDING_THREAD_NAME =
+        "org.apache.hadoop.fs.FileSystem$Statistics$StatisticsDataReferenceCleaner";
+
+    @Override
+    public boolean reject(Thread t) {
+        return t.getName().equalsIgnoreCase(OFFENDING_THREAD_NAME);
```
Comment (rjernst) [inline plugins/repository-hdfs/src/test/java/org/elasticsearch/repositories/hdfs/HdfsClientThreadLeakFilter.java]: Why is the ignore case needed?

File: plugins/repository-hdfs/build.gradle [hunk 5]
```diff
@@ -32,7 +32,7 @@ esplugin {
 apply plugin: 'elasticsearch.vagrantsupport'
 
 versions << [
-  'hadoop2': '2.7.1'
+  'hadoop2': '2.8.1'
```
Comment (rjernst) [inline plugins/repository-hdfs/build.gradle]: Did no dependency versions change? Just double checking; remember we don't use transitive dependencies, so upgrading means inspecting the new version's deps compared to what we currently pull in.

File: test/fixtures/hdfs-fixture/src/main/java/hdfs/MiniHDFS.java [hunk 6]
```diff
@@ -83,6 +85,8 @@ public static void main(String[] args) throws Exception {
             cfg.set(DFSConfigKeys.DFS_NAMENODE_KEYTAB_FILE_KEY, keytabFile);
             cfg.set(DFSConfigKeys.DFS_DATANODE_KEYTAB_FILE_KEY, keytabFile);
             cfg.set(DFSConfigKeys.DFS_NAMENODE_ACLS_ENABLED_KEY, "true");
+            // dfs.block.access.token.enable
```
Comment (rjernst) [inline test/fixtures/hdfs-fixture/src/main/java/hdfs/MiniHDFS.java]: nit: this comment line doesn't seem to add anything meaningful

**Review by rjernst — APPROVED** [review]
LGTM. The only major issue is whether any deps need updating.

File: plugins/repository-hdfs/build.gradle (hunk 5 above)
Comment (jbaiera) [inline plugins/repository-hdfs/build.gradle]: Htrace is the only new dependency with a new version

File: plugins/repository-hdfs/src/test/java/org/elasticsearch/repositories/hdfs/HdfsClientThreadLeakFilter.java (hunk 4 above)
Comment (jbaiera) [inline plugins/repository-hdfs/src/test/java/org/elasticsearch/repositories/hdfs/HdfsClientThreadLeakFilter.java]: Old habit of mine. I'll fix.

**Review by jasontedor — APPROVED** [review]
LGTM.

**Pull request merged**